cpp/src/arrow/dataset/file_orc.cc

```diff
@@ -0,0 +1,132 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/dataset/file_orc.h"
+
+#include <memory>
+
+#include "arrow/dataset/dataset_internal.h"
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/scanner.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+using internal::checked_pointer_cast;
+
+namespace dataset {
+
+static inline Result<std::unique_ptr<arrow::adapters::orc::ORCFileReader>> OpenReader(
+    const FileSource& source,
+    const std::shared_ptr<ScanOptions>& scan_options = nullptr) {
+  ARROW_ASSIGN_OR_RAISE(auto input, source.Open());
+
+  arrow::MemoryPool* pool;
+  if (scan_options) {
+    pool = scan_options->pool;
+  } else {
+    pool = default_memory_pool();
+  }
+
+  std::unique_ptr<arrow::adapters::orc::ORCFileReader> reader;
+  auto status =
+      arrow::adapters::orc::ORCFileReader::Open(std::move(input), pool, &reader);
+  if (!status.ok()) {
+    return status.WithMessage("Could not open ORC input source '", source.path(),
+                              "': ", status.message());
+  }
+  return reader;
+}
+
+/// \brief A ScanTask backed by an ORC file.
+class OrcScanTask : public ScanTask {
+ public:
+  OrcScanTask(std::shared_ptr<FileFragment> fragment,
+              std::shared_ptr<ScanOptions> options)
+      : ScanTask(std::move(options), fragment), source_(fragment->source()) {}
+
+  Result<RecordBatchIterator> Execute() override {
+    ARROW_ASSIGN_OR_RAISE(auto reader, OpenReader(source_));
+    std::shared_ptr<arrow::RecordBatchReader> batch_reader;
+    // TODO determine included fields from options_->MaterializedFields() to
+    // optimize the column selection (see _column_index_lookup in python
+    // orc.py for custom logic)
+    // std::vector<int> included_fields;
+    RETURN_NOT_OK(reader->NextStripeReader(options_->batch_size, &batch_reader));
```

Comment:
Hmm, IIUC, this will only read one stripe from the ORC file, but an ORC file may well have several stripes. @iajoiner is my interpretation correct?

Reply:
Ah, I was assuming this would iteratively read the "next" stripe on each call of the iterator, but I might have misunderstood that. 

We also have https://issues.apache.org/jira/browse/ARROW-13571 to control the stripe size when writing, which will also make it easier to test this with small datasets.

Reply:
OK, indeed, doing a quick test with writing a file with 10 million rows (so it has 2 stripes), and reading it back with the datasets API of this PR, it only reads around 8 million rows .. 
So will have to look for a different way.

Reply:
I pushed an update, and now manually iteratively read the next stripe with `ReadStripe` (I followed the approach of a ScanTaskIterator from file_ipc.cc)

Reply:
@pitrou  I'm really late here. For the record the answer is "Yes".